lib/Sema/SemanticResolver.cpp

```diff
@@ -93,6 +93,8 @@ bool SemanticResolver::runLazy(
     if (astContext_.getDebugInfoSetting() != DebugInfoSetting::ALL) {
       semInfo->bindingTableScope.reset();
     }
+
+    semCtx_.clearPromotedDecls();
```

Comment:
I think this is the right fix, good catch.

Let's also put a comment so we don't forget why: the `promotedDecls` is used immediately, and keyed off the `IdentifierNode`, which is freed between invocations of lazy compilation. We need to remove references in `SemContext` when freeing the AST, which happens at the end of each invocation of lazy compilation.